File: Graphs/834. Sum of Distances in Tree.py
class Solution:
    def sumOfDistancesInTree(self, n: int, edges: List[List[int]]) -> List[int]:
        graph = collections.defaultdict(set)
        res = [0] * n
        count = [1] * n
        for i, j in edges:
            graph[i].add(j)
            graph[j].add(i)

        def dfs(currentNode, parent): # postorder traversal 
            for kid in graph[currentNode]:
                if kid != parent:
                    dfs(kid, currentNode)
                    count[currentNode] += count[kid] # the number of nodes in every subtree
                    res[currentNode] += res[kid] + count[kid] # the distance of every node from its parent node 

        def dfs2(currentNode, parent): # preorder traversal 
            for kid in graph[currentNode]:
                if kid != parent:
                    res[kid] = res[currentNode] - count[kid] + n - count[kid]  # the result of a tree is equals to the distances of children nodes of the currentNode - count of the children node + n - count[child] because by moving downwards, the total distance of the childSubtree is discounted by 1, +(n - count[child]) because anything else other than that gets incremented by 1 
                    
                    dfs2(kid, currentNode)

        dfs(0, -1)
        dfs2(0, -1)

        return res
        

File: Binary Tree/834. Sum of Distances in Tree.cpp
class Solution {
public:
    vector<int>cnt;
    vector<int>res;
    int N;
    vector<vector<int>>g;
    void dfs1(int cur = 0, int p = -1) {
        cnt[cur] = 1;
        for (auto x : g[cur]) {
            if (x != p) {
                dfs1(x, cur);
                cnt[cur] += cnt[x];
                res[cur] += res[x] + cnt[x];
            }
        }
    }

    void dfs2(int cur = 0 , int p = -1) {

        for (auto x : g[cur]) {
            if (x != p) {
                res[x] = res[cur] - cnt[x] + N - cnt[x];
                dfs2(x, cur);
            }
        }
    }
    vector<int> sumOfDistancesInTree(int n, vector<vector<int>>& edges) {
        res.resize(n);
        g.resize(n);
        cnt.resize(n);
        for (auto& x : edges) {
            int a = x[0], b = x[1];
            g[a].push_back(b);
            g[b].push_back(a);
        }

        N= n;

        dfs1();
        dfs2();

        return res;
    }
};